File: 03-stream-processing/python/exercise4/data_generator.py
import json
import os
import time
from typing import Dict

import pendulum as pendulum
from pendulum import DateTime
from pyspark.sql import SparkSession

print('Creating topics')
topic_name = 'data_for_backpressure'
os.system(
    f'docker exec p4_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --delete')
os.system(
    f'docker exec p4_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic {topic_name} --create --partitions 2')


def row_to_write(id: int, row_value: str, generation_time: DateTime) -> Dict[str, str]:
    return {'value': json.dumps({'id': id, 'row_value': row_value,
                       'generation_time': generation_time.to_iso8601_string()})}


spark = SparkSession.builder.master("local[*]") \
    .config('spark.jars.packages', 'org.apache.spark:spark-sql-kafka-0-10_2.12:3.4.0').getOrCreate()

generation_day = pendulum.now('UTC').subtract(days=10)
while pendulum.now('UTC') > generation_day:
    print('Sending data')
    visits_to_generate = spark.createDataFrame(data=[
        row_to_write(1, "A", generation_day.add(seconds=1)), row_to_write(1, "E", generation_day.add(seconds=2)),
        row_to_write(1, "I", generation_day.add(seconds=3)), row_to_write(1, "O", generation_day.add(seconds=4)),
        row_to_write(1, "U", generation_day.add(seconds=5)), row_to_write(1, "AA", generation_day.add(seconds=6)),
        row_to_write(1, "EE", generation_day.add(seconds=7))
    ], schema='value STRING')
    visits_to_generate.write.format("kafka")\
        .option("topic", topic_name).option("kafka.bootstrap.servers", "localhost:29092")\
        .save()

    generation_day = generation_day.add(minutes=1)
    time.sleep(5)

File: 03-stream-processing/scala/src/main/scala/com/becomedataengineer/exercise4/DataGenerator.scala
package com.becomedataengineer.exercise4

import org.apache.spark.sql.{SparkSession, functions}

import java.time.{LocalDateTime, ZoneId}
import scala.util.Try
import scala.sys.process._

object DataGenerator {

  def main(args: Array[String]): Unit = {
    println("Preparing topics")
    val topicName = "data_for_backpressure"
    Try {
      val cmdDelete = s"docker exec p4_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic ${topicName} --delete"
      val outputDelete = cmdDelete.!!
    }
    val cmdCreate = s"docker exec p4_kafka kafka-topics.sh --bootstrap-server localhost:29092 --topic ${topicName} --create --partitions 2"
    val outputCreate = cmdCreate.!!

    val sparkSession = SparkSession.builder().master("local[*]")
      .getOrCreate()
    import sparkSession.implicits._
    val utc = ZoneId.of("UTC")
    val firstDayToGenerate = LocalDateTime.now(utc).minusDays(10)
    var generationTime = firstDayToGenerate
    while (generationTime.isBefore(LocalDateTime.now(utc))) {
      println("Sending data")
      Seq(
        (1, "A", generationTime.plusSeconds(1)), (1, "E", generationTime.plusSeconds(2)),
        (1, "I", generationTime.plusSeconds(3)), (1, "O", generationTime.plusSeconds(4)),
        (1, "U", generationTime.plusSeconds(4)), (1, "AA", generationTime.plusSeconds(5)),
        (1, "EE", generationTime.plusSeconds(6))
        ).toDF("id", "row_value", "generation_time")
        .withColumn("value", functions.to_json(functions.struct("id", "row_value", "generation_time")))
        .select("value")
        .write.format("kafka")
        .option("topic", topicName)
        .option("kafka.bootstrap.servers", "localhost:29092")
        .save()

      generationTime = generationTime.plusMinutes(1L)
      Thread.sleep(5000L)
    }
  }
}